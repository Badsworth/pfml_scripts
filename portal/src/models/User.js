import BaseModel from "./BaseModel";

class User extends BaseModel {
  get defaults() {
    return {
      user_id: null,
      auth_id: null,
      email_address: null,
      first_name: null,
      middle_name: null,
      last_name: null,
      date_of_birth: null,
      has_state_id: null,
      state_id: null,
      status: null,
      ssn_or_itin: null,
    };
  }
}

export default User;
