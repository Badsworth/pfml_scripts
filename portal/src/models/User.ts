/* eslint sort-keys: ["error", "asc"] */
export const PhoneType = {
  cell: "Cell",
  phone: "Phone",
} as const;

export const MFAPreference = {
  optOut: "Opt Out",
  sms: "SMS",
} as const;

class User {
  auth_id: string;
  consented_to_data_sharing: boolean;
  email_address: string;
  roles: UserRole[] = [];
  user_id: string;
  mfa_phone_number: {
    int_code: string | null;
    phone_number: string | null;
    phone_type: typeof PhoneType[keyof typeof PhoneType] | null;
  } | null = null;

  mfa_delivery_preference:
    | typeof MFAPreference[keyof typeof MFAPreference]
    | null = null;

  user_leave_administrators: UserLeaveAdministrator[] = [];

  constructor(attrs: Partial<User>) {
    Object.assign(this, attrs);
  }

  get hasEmployerRole() {
    return this.roles.some(
      (userRole) => userRole.role_description === RoleDescription.employer
    );
  }

  /**
   * Determines whether user_leave_administrators has only unverified employers
   */
  get hasOnlyUnverifiedEmployers(): boolean {
    return this.verifiedEmployers.length === 0;
  }

  /**
   * Determines whether user_leave_administrators has a verifiable employer
   */
  get hasVerifiableEmployer(): boolean {
    return this.user_leave_administrators.some((employer) =>
      this.isVerifiableEmployer(employer)
    );
  }

  /**
   * Returns list of verified employers
   */
  get verifiedEmployers(): UserLeaveAdministrator[] {
    return this.user_leave_administrators.filter(
      (employer) => employer.verified === true
    );
  }

  /**
   * Determines whether an employer is verifiable (unverified and has verification data)
   */
  isVerifiableEmployer(employer: UserLeaveAdministrator): boolean {
    return !employer.verified && employer.has_verification_data;
  }

  /**
   * Determines whether an employer is registered in FINEOS
   */
  isEmployerIdRegisteredInFineos(employerId: string | null): boolean {
    return this.user_leave_administrators.some(
      (employer) =>
        employerId === employer.employer_id && employer.has_fineos_registration
    );
  }
}

export class UserRole {
  role_description: typeof RoleDescription[keyof typeof RoleDescription];
  role_id: number;

  constructor(attrs: Partial<UserRole>) {
    Object.assign(this, attrs);
  }
}

/**
 * Enums for UserRole's `role_description` field
 * @enum {string}
 */
export const RoleDescription = {
  claimant: "Claimant",
  employer: "Employer",
} as const;

export class UserLeaveAdministrator {
  employer_dba: string;
  employer_fein: string;
  employer_id: string;
  has_fineos_registration: boolean;
  has_verification_data: boolean;
  verified: boolean;

  constructor(attrs: Partial<UserLeaveAdministrator>) {
    Object.assign(this, attrs);
  }
}

export default User;
