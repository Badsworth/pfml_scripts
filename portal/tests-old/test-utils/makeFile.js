/**
 * Make a File instance
 * @see https://developer.mozilla.org/en-US/docs/Web/API/File
 * @param {object} attrs File attributes
 * @param {string} attrs.name The file's name
 * @param {string} attrs.type The file's MIME type
 * @returns {File}
 */
export const makeFile = (attrs = {}) => {
  const { name, type } = Object.assign(
    {
      name: "file.pdf",
      type: "application/pdf",
    },
    attrs
  );

  return new File([], name, { type });
};

/**
 * Create an HTML <input> for a test's event.target
 * @param {object} attrs - HTML attributes for the input
 * @returns {HTMLInputElement}
 */
export function createInputElement(attrs) {
  const input = document.createElement("input");

  Object.entries(attrs).forEach(([attributeName, value]) => {
    input.setAttribute(attributeName, value);
  });

  if (attrs.checked === false) input.checked = false;

  return input;
}
