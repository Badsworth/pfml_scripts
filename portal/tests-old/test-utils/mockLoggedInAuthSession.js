import { Auth } from "@aws-amplify/auth";

/**
 * Mock the Cognito authentication session so a test user appears logged in
 */
const mockLoggedInAuthSession = () => {
  const accessTokenJwt =
    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiQnVkIn0.YDRecdsqG_plEwM0H8rK7t2z0R3XRNESJB5ZXk-FRN8";

  jest.spyOn(Auth, "currentSession").mockImplementation(() =>
    Promise.resolve({
      accessToken: { jwtToken: accessTokenJwt },
    })
  );
};

export default mockLoggedInAuthSession;
