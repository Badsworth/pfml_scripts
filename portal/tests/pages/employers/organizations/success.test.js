import User, { UserLeaveAdministrator } from "../../../../src/models/User";
import Success from "../../../../src/pages/employers/organizations/success";
import { renderPage } from "../../../test-utils";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const setup = (props = {}) => {
  let goToSpy;
  const utils = renderPage(
    Success,
    {
      addCustomSetup: (appLogic) => {
        appLogic.users.user = new User({
          consented_to_data_sharing: true,
          user_leave_administrators: [
            new UserLeaveAdministrator({
              employer_dba: "Company Name",
              employer_fein: "12-3456789",
              employer_id: "mock_employer_id",
              verified: false,
            }),
          ],
        });
        goToSpy = jest.spyOn(appLogic.portalFlow, "goTo");
      },
    },
    {
      query: { employer_id: "mock_employer_id", next: "" },
      ...props,
    }
  );
  return { goToSpy, ...utils };
};
describe("Success", () => {
  it("renders the page", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("navigates to Organizations page when clicking 'Continue' button", () => {
    const { goToSpy } = setup();
    userEvent.click(screen.getByRole("button", { name: "Continue" }));
    expect(goToSpy).toHaveBeenCalledWith("/employers/organizations");
  });

  it("navigates to Organizations page by default if query param is invalid", () => {
    const queryWithoutNextParam = {
      employer_id: "mock_employer_id",
      next: "",
    };
    const { goToSpy } = setup({
      query: queryWithoutNextParam,
    });
    userEvent.click(screen.getByRole("button", { name: "Continue" }));
    expect(goToSpy).toHaveBeenCalledWith("/employers/organizations");
  });

  it("navigates to New Application page based on next param", () => {
    const queryWithNextParam = {
      employer_id: "mock_employer_id",
      next: "/employers/applications/new-application/?absence_id=mock_absence_id",
    };

    const { goToSpy } = setup({
      query: queryWithNextParam,
    });
    userEvent.click(screen.getByRole("button", { name: "Continue" }));
    expect(goToSpy).toHaveBeenCalledWith(
      "/employers/applications/new-application/?absence_id=mock_absence_id"
    );
  });
});
