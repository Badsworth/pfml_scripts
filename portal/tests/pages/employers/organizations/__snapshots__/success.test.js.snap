// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Success renders the page 1`] = `
<Fragment>
  <Title>
    Thanks for verifying your paid leave contributions
  </Title>
  <p>
    <Trans
      i18nKey="pages.employersOrganizationsSuccess.companyNameLabel"
      tOptions={
        Object {
          "company": "Company Name",
        }
      }
    />
    <br />
    <Trans
      i18nKey="pages.employersOrganizationsSuccess.employerIdNumberLabel"
      tOptions={
        Object {
          "ein": "11-111111",
        }
      }
    />
  </p>
  <p>
    <Trans
      components={
        Object {
          "learn-more-link": <a
            href="https://www.mass.gov/how-to/creating-an-employer-account-to-review-paid-family-and-medical-leave-pfml-applications"
            rel="noopener"
            target="_blank"
          />,
        }
      }
      i18nKey="pages.employersOrganizationsSuccess.instructions"
    />
  </p>
  <Button
    onClick={[Function]}
  >
    Continue
  </Button>
</Fragment>
`;

exports[`Success renders the page 2`] = `
Array [
  <strong
    key="strong-0"
  >
    Organization:
  </strong>,
  " Company Name",
]
`;

exports[`Success renders the page 3`] = `
Array [
  <strong
    key="strong-0"
  >
    Employer ID number (EIN):
  </strong>,
  " 11-111111",
]
`;

exports[`Success renders the page 4`] = `
Array [
  "Your account has been verified. It may take up to 15 minutes for our systems to update so that you can log in and review applications. If anyone else on your team needs to review applications, theyâ€™ll also need to complete the ",
  <a
    href="https://www.mass.gov/how-to/creating-an-employer-account-to-review-paid-family-and-medical-leave-pfml-applications"
    key="1"
    rel="noopener"
    target="_blank"
  >
    verification process
  </a>,
  ".",
]
`;
